### ISSUE TITLE
the loadBalance() method does not work for the Norne deck if MPI is enabled

### ISSUE THREAD

**Issue by andlaus** [OPM/dune-cornerpoint#131]
this is even for single process runs (i.e., the `loadBalance()` method is supposed to be a no-op in this case):

the test program:
```c++
#include "config.h"
#include <dune/grid/CpGrid.hpp>
#include <opm/parser/eclipse/Parser/Parser.hpp>
#include <opm/parser/eclipse/Deck/Deck.hpp>
#include <dune/common/parallel/mpihelper.hh>

int main(int argc, char **argv)
{
    Dune::MPIHelper::instance(argc, argv);
    Opm::ParserPtr parser(new Opm::Parser);
    Opm::DeckConstPtr deck(parser->parseFile(argv[1]));

    Dune::CpGrid grid;
    grid.processEclipseFormat(deck, /*zTolerance=*/0, /*isPeriodic=*/false);
    grid.loadBalance();
    std::cout << "hello, world!\n";
    return 0;
}
```

result on dune master with MPI disabled:
```
./test_prog /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
CpGrid::scatterGrid() is non-trivial only with MPI support and if the target Dune platform is sufficiently recent.
hello, world!
```
(which is a bit confusing, but it does not fall on its nose.)

result with MPI support enabled:
```
./test_prog /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
terminate called after throwing an instance of 'std::runtime_error'
  what():  [/home/and/src/dune-cornerpoint/dune/grid/cpgrid/CpGridData.cpp:536] rank for cell is too big
[heuristix:32579] *** Process received signal ***
[heuristix:32579] Signal: Aborted (6)
[heuristix:32579] Signal code:  (-6)
[heuristix:32579] [ 0] /lib64/libpthread.so.0(+0xf890) [0x7f9f1ce1d890]
[heuristix:32579] [ 1] /lib64/libc.so.6(gsignal+0x37) [0x7f9f1c372187]
[heuristix:32579] [ 2] /lib64/libc.so.6(abort+0x118) [0x7f9f1c373538]
[heuristix:32579] [ 3] /usr/lib64/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x175) [0x7f9f1c960975]
[heuristix:32579] [ 4] /usr/lib64/libstdc++.so.6(+0x639f6) [0x7f9f1c95e9f6]
[heuristix:32579] [ 5] /usr/lib64/libstdc++.so.6(+0x63a41) [0x7f9f1c95ea41]
[heuristix:32579] [ 6] /usr/lib64/libstdc++.so.6(+0x63c58) [0x7f9f1c95ec58]
[heuristix:32579] [ 7] ./bin/tutorial1(_ZN4Dune6cpgrid10CpGridData20distributeGlobalGridERKNS_6CpGridERKS1_RKSt6vectorIiSaIiEEi+0x1f5) [0x592835]
[heuristix:32579] [ 8] ./bin/tutorial1(_ZN4Dune6CpGrid11scatterGridEi+0x55f) [0x55553f]
[heuristix:32579] [ 9] ./bin/tutorial1(_ZN4Dune6CpGrid11loadBalanceEi+0x1b) [0x54c22b]
[heuristix:32579] [10] ./bin/tutorial1(main+0x1d8) [0x54b178]
[heuristix:32579] [11] /lib64/libc.so.6(__libc_start_main+0xf5) [0x7f9f1c35eb05]
[heuristix:32579] [12] ./bin/tutorial1() [0x54aec3]
[heuristix:32579] *** End of error message ***
Aborted
```

gdb says that this happens somewhere in the call to `loadBalance()`...

**Comment by bska** [comment]
>
```
./test_prog /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
```

Just out of curiosity: Do you get the same failures in the MPI case if you launch the utility using
```sh
$ mpirun -np 1 ./test_prog /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
```

**Comment by andlaus** [comment]
yep, same story:

```
mpirun -np 1 ./test_prog /home/and/src/opm-data/norne/NORNE_ATW2013.DATA
terminate called after throwing an instance of 'std::runtime_error'
  what():  [/home/and/src/dune-cornerpoint/dune/grid/cpgrid/CpGridData.cpp:536] rank for cell is too big
[heuristix:05644] *** Process received signal ***
[heuristix:05644] Signal: Aborted (6)
[heuristix:05644] Signal code:  (-6)
[heuristix:05644] [ 0] /lib64/libpthread.so.0(+0xf890) [0x7fb81b8b9890]
[heuristix:05644] [ 1] /lib64/libc.so.6(gsignal+0x37) [0x7fb81ae0e187]
[heuristix:05644] [ 2] /lib64/libc.so.6(abort+0x118) [0x7fb81ae0f538]
[heuristix:05644] [ 3] /usr/lib64/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x175) [0x7fb81b3fc975]
[heuristix:05644] [ 4] /usr/lib64/libstdc++.so.6(+0x639f6) [0x7fb81b3fa9f6]
[heuristix:05644] [ 5] /usr/lib64/libstdc++.so.6(+0x63a41) [0x7fb81b3faa41]
[heuristix:05644] [ 6] /usr/lib64/libstdc++.so.6(+0x63c58) [0x7fb81b3fac58]
[heuristix:05644] [ 7] ./bin/tutorial1(_ZN4Dune6cpgrid10CpGridData20distributeGlobalGridERKNS_6CpGridERKS1_RKSt6vectorIiSaIiEEi+0x1f5) [0x592835]
[heuristix:05644] [ 8] ./bin/tutorial1(_ZN4Dune6CpGrid11scatterGridEi+0x55f) [0x55553f]
[heuristix:05644] [ 9] ./bin/tutorial1(_ZN4Dune6CpGrid11loadBalanceEi+0x1b) [0x54c22b]
[heuristix:05644] [10] ./bin/tutorial1(main+0x1d8) [0x54b178]
[heuristix:05644] [11] /lib64/libc.so.6(__libc_start_main+0xf5) [0x7fb81adfab05]
[heuristix:05644] [12] ./bin/tutorial1() [0x54aec3]
[heuristix:05644] *** End of error message ***
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 5644 on node heuristix exited on signal 6 (Aborted).
--------------------------------------------------------------------------
```

**Comment by blattms** [comment]
I seems that I have misinterpreted the non-existing documentation in
GridPartitioning.{hc}pp. I always assumed that the number of
partitions specified serves as an upper limit. Now I see that in
```ensureConnectedPartitons``` new partitions might be created. It
seems like that for each partition where not all cells can be
connected via a path that stays within the partition the isolated cell
complexes will created additional partitions.

Long story short, the number of partitions might be larger than
initially requested. In this case we request one partition but more
than one are actually created.

Maybe one of the original authors can comment on the original use case
of the ```partition``` method?

**Comment by atgeirr** [comment]
It was designed with multi-scale methods in mind, but has not been put to that use.

I do think the documentation is adequate however, it explicitly says that it can produce fewer or more partitions depending on the grid, than asked for in the logical cartesian sense. What is it you find missing?

Also, if you get more partitions than one when you specify a (1, 1, 1) split, it means that the grid is not connected. This may or may not be a problem in itself. I think we can handle this in Flow, but not all codes may accept it.

**Comment by bska** [comment]
> Maybe one of the original authors can comment on the original use case of the `partition` method?

That would be me.  That method was used in a (now defunct) implementation of the "Multiscale Mixed Finite-Element (MsMFE)" method.  When used in that context, it is absolutely critical that a block (a subset of existing cells) be connected through a neighbourship relation.  Blocks that don't consist of connected sets of cells are split into their constituent connected components (depth-first graph search) so it is definitely possible that the number of blocks increases as a result of partition processing.

To be fair, this is (sort of) documented in the synopsis of `partition()` in the comment
```C++
/// Partition a CpGrid based on (ijk) coordinates, with splitting to ensure that each partition is connected.
```
The important part is the **with splitting to ensure...**.

**Comment by atgeirr** [comment]
Is this resolved now and can be closed?

**Comment by andlaus** [comment]
it now works. thanks
